Micronaut Upgrade to version: 3.8.2 giving java.lang.VerifyError: Bad type on operand stack

### Discussed in https://github.com/micronaut-projects/micronaut-core/discussions/8653

<div type='discussions-op-text'>

<sup>Originally posted by **npathai** January 23, 2023</sup>
When trying to upgrade from Micronaut version: 3.3.3 to version 3.8.2, I am getting following error.

Micronaut Version: 3.8.2 (But the issue also occurs with >= 3.4.0)
Java version: 11
Micronaut Data Version: 3.9.4


```
Caused by: java.lang.VerifyError: Bad type on operand stack

  Location:
    consumerdetails/$ConsumerDetails$Introspection.dispatch(ILjava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object; @28: invokevirtual
  Reason:
    Type 'consumerdetails/$ConsumerDetails$Introspection' (current frame, stack[1]) is not assignable to 'io/micronaut/context/AbstractExecutableMethodsDefinition'
  Current Frame:
    bci: @28
    flags: { }
    locals: { 'consumerdetails/$ConsumerDetails$Introspection', integer, 'java/lang/Object', '[Ljava/lang/Object;' }
    stack: { 'java/lang/Object', 'consumerdetails/$ConsumerDetails$Introspection', integer }
  Bytecode:
    0000000: 1baa 0000 0000 0035 0000 0014 0000 0015
    0000010: 0000 0017 0000 0026 2c2a 1014 b601 06b2
    0000020: 010c b801 1201 b02c 2a10 15b6 0106 b201
    0000030: 0cb8 0112 01b0 2a1b b600 fcbf          
  Stackmap Table:
    same_frame(@24)
    same_frame(@39)
    same_frame(@54)

	at consumerdetails.$ConsumerDetails$IntrospectionRef.load(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1693)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.core.beans.DefaultBeanIntrospector.findIntrospections(DefaultBeanIntrospector.java:67)
	at io.micronaut.core.beans.BeanIntrospector.findIntrospections(BeanIntrospector.java:91)
	at io.micronaut.configuration.hibernate.jpa.JpaConfiguration$EntityScanConfiguration.findEntities(JpaConfiguration.java:389)
	at io.micronaut.configuration.hibernate.jpa.conf.AbstractHibernateFactory.buildMetadata(AbstractHibernateFactory.java:78)
	at io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory.buildMetadata(SessionFactoryPerDataSourceFactory.java:88)
	at io.micronaut.configuration.hibernate.jpa.conf.$SessionFactoryPerDataSourceFactory$BuildMetadata2$Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractInitializableBeanDefinition.build(AbstractInitializableBeanDefinition.java:769)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:158)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2354)
```</div>

--- a/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/Auditable.java
+++ b/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/Auditable.java
@@ -1,0 +1,20 @@
+package io.micronaut.inject.visitor.beans;
+
+import io.micronaut.context.annotation.Executable;
+
+import java.time.Instant;
+
+public abstract class Auditable {
+
+    private Instant updatedAt;
+
+    @Executable(processOnStartup = true)
+    protected void beforeInsert() {
+        updatedAt = Instant.now();
+    }
+
+    public Instant getUpdatedAt() {
+        return updatedAt;
+    }
+}
+

--- a/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionGenericsSpec.groovy
+++ b/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionGenericsSpec.groovy
@@ -1,10 +1,13 @@
 package io.micronaut.inject.visitor.beans
 
 import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
+import io.micronaut.core.beans.BeanIntrospection
 import io.micronaut.core.type.GenericPlaceholder
+import spock.lang.Issue
 
 class BeanIntrospectionGenericsSpec
         extends AbstractTypeElementSpec {
+
     void "test generic placeholder for bean properties"() {
         given:
         def introspection = buildBeanIntrospection('test.Test', '''

--- a/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionSpec.groovy
+++ b/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionSpec.groovy
@@ -5,6 +5,7 @@ import com.fasterxml.jackson.annotation.JsonClassDescription
 import com.fasterxml.jackson.annotation.JsonCreator
 import com.fasterxml.jackson.annotation.JsonProperty
 import com.fasterxml.jackson.databind.ObjectMapper
+import groovy.transform.PackageScope
 import io.micronaut.annotation.processing.TypeElementVisitorProcessor
 import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
 import io.micronaut.annotation.processing.test.JavaParser
@@ -42,8 +43,44 @@ import javax.validation.constraints.Min
 import javax.validation.constraints.NotBlank
 import javax.validation.constraints.Size
 import java.lang.reflect.Field
+import java.time.Instant
 
 class BeanIntrospectionSpec extends AbstractTypeElementSpec {
+
+    @Issue("https://github.com/micronaut-projects/micronaut-core/issues/8657")
+    void "test executable method on abstract class with introspection"() {
+        when:
+        def introspection = buildBeanIntrospection('issue8657.Test', '''
+package issue8657;
+
+import io.micronaut.core.annotation.Introspected;
+import io.micronaut.context.annotation.Executable;
+import io.micronaut.inject.visitor.beans.Auditable;
+
+@Introspected
+class Test extends Auditable {
+    private String name;
+    public void setName(String name) {
+        this.name = name;
+    }
+    public String getName() {
+        return name;
+    }
+}
+
+''')
+
+        then:
+        introspection.beanMethods.size() == 1
+
+        when:
+        def bean = introspection.instantiate()
+        def method = introspection.beanMethods.first()
+        method.invoke(bean)
+
+        then:
+        bean.updatedAt != null
+    }
 
     void "test generics in arrays don't stack overflow"() {
         given:
@@ -4288,3 +4325,5 @@ public record Foo(String name, String isSurname, boolean contains, Boolean purge
         }
     }
 }
+
+

--- a/inject/src/main/java/io/micronaut/context/AbstractExecutableMethodsDefinition.java
+++ b/inject/src/main/java/io/micronaut/context/AbstractExecutableMethodsDefinition.java
@@ -44,7 +44,6 @@ public abstract class AbstractExecutableMethodsDefinition<T> implements Executab
     private final DispatchedExecutableMethod<T, ?>[] executableMethods;
     private Environment environment;
     private List<DispatchedExecutableMethod<T, ?>> executableMethodsList;
-    private Method[] reflectiveMethods;
 
     protected AbstractExecutableMethodsDefinition(MethodReference[] methodsReferences) {
         this.methodsReferences = methodsReferences;
@@ -161,17 +160,11 @@ public abstract class AbstractExecutableMethodsDefinition<T> implements Executab
     // this logic must allow reflection
     @SuppressWarnings("java:S3011")
     protected final Method getAccessibleTargetMethodByIndex(int index) {
-        if (reflectiveMethods == null) {
-            reflectiveMethods = new Method[methodsReferences.length];
+        Method method = getTargetMethodByIndex(index);
+        if (ClassUtils.REFLECTION_LOGGER.isDebugEnabled()) {
+            ClassUtils.REFLECTION_LOGGER.debug("Reflectively accessing method {} of type {}", method, method.getDeclaringClass());
         }
-        Method method = reflectiveMethods[index];
-        if (method == null) {
-            method = getTargetMethodByIndex(index);
-            if (ClassUtils.REFLECTION_LOGGER.isDebugEnabled()) {
-                ClassUtils.REFLECTION_LOGGER.debug("Reflectively accessing method {} of type {}", method, method.getDeclaringClass());
-            }
-            method.setAccessible(true);
-        }
+        method.setAccessible(true);
         return method;
     }
 

--- a/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
+++ b/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
@@ -120,6 +120,35 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements BeanI
     }
 
     /**
+     * Find {@link Method} representation at the method by index. Used by {@link ExecutableMethod#getTargetMethod()}.
+     *
+     * @param index The index
+     * @return The method
+     */
+    @UsedByGeneratedCode
+    @Internal
+    protected abstract Method getTargetMethodByIndex(int index);
+
+    /**
+     * Find {@link Method} representation at the method by index. Used by {@link ExecutableMethod#getTargetMethod()}.
+     *
+     * @param index The index
+     * @return The method
+     * @since 3.8.5
+     */
+    @UsedByGeneratedCode
+    // this logic must allow reflection
+    @SuppressWarnings("java:S3011")
+    protected final Method getAccessibleTargetMethodByIndex(int index) {
+        Method method = getTargetMethodByIndex(index);
+        if (ClassUtils.REFLECTION_LOGGER.isDebugEnabled()) {
+            ClassUtils.REFLECTION_LOGGER.debug("Reflectively accessing method {} of type {}", method, method.getDeclaringClass());
+        }
+        method.setAccessible(true);
+        return method;
+    }
+
+    /**
      * Triggers the invocation of the method at index.
      *
      * @param index  The method index
@@ -577,7 +606,7 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements BeanI
             if (ClassUtils.REFLECTION_LOGGER.isWarnEnabled()) {
                 ClassUtils.REFLECTION_LOGGER.warn("Using getTargetMethod for method {} on type {} requires the use of reflection. GraalVM configuration necessary", getName(), getDeclaringType());
             }
-            return ReflectionUtils.getRequiredMethod(getDeclaringType(), getMethodName(), getArgumentTypes());
+            return getTargetMethodByIndex(ref.methodIndex);
         }
 
         @Override

--- a/inject/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
+++ b/inject/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
@@ -136,7 +136,7 @@ final class BeanIntrospectionWriter extends AbstractAnnotationMetadataWriter {
         this.introspectionName = computeIntrospectionName(name);
         this.introspectionType = getTypeReferenceForName(introspectionName);
         this.beanType = getTypeReferenceForName(name);
-        this.dispatchWriter = new DispatchWriter(introspectionType);
+        this.dispatchWriter = new DispatchWriter(introspectionType, Type.getType(AbstractInitializableBeanIntrospection.class));
     }
 
     /**
@@ -618,6 +618,7 @@ final class BeanIntrospectionWriter extends AbstractAnnotationMetadataWriter {
 
         dispatchWriter.buildDispatchOneMethod(classWriter);
         dispatchWriter.buildDispatchMethod(classWriter);
+        dispatchWriter.buildGetTargetMethodByIndex(classWriter);
         buildPropertyIndexOfMethod(classWriter);
         buildFindIndexedProperty(classWriter);
         buildGetIndexedProperties(classWriter);

--- a/inject/src/main/java/io/micronaut/inject/writer/DispatchWriter.java
+++ b/inject/src/main/java/io/micronaut/inject/writer/DispatchWriter.java
@@ -71,11 +71,19 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
 
     private final List<DispatchTarget> dispatchTargets = new ArrayList<>();
     private final Type thisType;
+
+    private final Type dispatchSuperType;
+
     private boolean hasInterceptedMethod;
 
     public DispatchWriter(Type thisType) {
+        this(thisType, ExecutableMethodsDefinitionWriter.SUPER_TYPE);
+    }
+
+    public DispatchWriter(Type thisType, Type dispatchSuperType) {
         super();
         this.thisType = thisType;
+        this.dispatchSuperType = dispatchSuperType;
     }
 
     /**
@@ -118,7 +126,7 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
      * @return the target index
      */
     public int addMethod(TypedElement declaringType, MethodElement methodElement, boolean useOneDispatch) {
-        return addDispatchTarget(new MethodDispatchTarget(declaringType, methodElement, useOneDispatch, !useOneDispatch));
+        return addDispatchTarget(new MethodDispatchTarget(dispatchSuperType, declaringType, methodElement, useOneDispatch, !useOneDispatch));
     }
 
     /**
@@ -136,6 +144,7 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
                                     String interceptedProxyBridgeMethodName) {
         hasInterceptedMethod = true;
         return addDispatchTarget(new InterceptableMethodDispatchTarget(
+                dispatchSuperType,
                 declaringType,
                 methodElement,
                 interceptedProxyClassName,
@@ -466,15 +475,17 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
     @Internal
     @SuppressWarnings("FinalClass")
     public static class MethodDispatchTarget implements DispatchTarget {
+        final Type dispatchSuperType;
         final TypedElement declaringType;
         final MethodElement methodElement;
         final boolean oneDispatch;
         final boolean multiDispatch;
 
-        private MethodDispatchTarget(TypedElement declaringType,
+        private MethodDispatchTarget(Type dispatchSuperType, TypedElement declaringType,
                                      MethodElement methodElement,
                                      boolean oneDispatch,
                                      boolean multiDispatch) {
+            this.dispatchSuperType = dispatchSuperType;
             this.declaringType = declaringType;
             this.methodElement = methodElement;
             this.oneDispatch = oneDispatch;
@@ -519,7 +530,7 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
                 }
                 writer.loadThis();
                 writer.push(methodIndex);
-                writer.invokeVirtual(ExecutableMethodsDefinitionWriter.SUPER_TYPE, GET_ACCESSIBLE_TARGET_METHOD);
+                writer.invokeVirtual(dispatchSuperType, GET_ACCESSIBLE_TARGET_METHOD);
                 if (hasArgs) {
                     writer.loadArg(2);
                 } else {
@@ -598,12 +609,13 @@ public final class DispatchWriter extends AbstractClassFileWriter implements Opc
         final String interceptedProxyBridgeMethodName;
         final Type thisType;
 
-        private InterceptableMethodDispatchTarget(TypedElement declaringType,
+        private InterceptableMethodDispatchTarget(Type dispatchSuperType,
+                                                  TypedElement declaringType,
                                                   MethodElement methodElement,
                                                   String interceptedProxyClassName,
                                                   String interceptedProxyBridgeMethodName,
                                                   Type thisType) {
-            super(declaringType, methodElement, false, true);
+            super(dispatchSuperType, declaringType, methodElement, false, true);
             this.interceptedProxyClassName = interceptedProxyClassName;
             this.interceptedProxyBridgeMethodName = interceptedProxyBridgeMethodName;
             this.thisType = thisType;

